graylog2-server/src/test/java/org/graylog2/shared/buffers/processors/MessageTimestampTest.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.shared.buffers.processors;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.eventbus.EventBus;
+import jakarta.inject.Provider;
+import org.graylog.failure.FailureSubmissionService;
+import org.graylog2.buffers.OutputBuffer;
+import org.graylog2.messageprocessors.OrderedMessageProcessors;
+import org.graylog2.plugin.InstantMillisProvider;
+import org.graylog2.plugin.Message;
+import org.graylog2.plugin.MessageFactory;
+import org.graylog2.plugin.TestMessageFactory;
+import org.graylog2.plugin.buffers.MessageEvent;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.graylog2.plugin.streams.Stream;
+import org.graylog2.streams.StreamMetrics;
+import org.graylog2.system.processing.ProcessingStatusRecorder;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeUtils;
+import org.joda.time.DateTimeZone;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+import java.time.Duration;
+import java.util.HashSet;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.when;
+
+class MessageTimestampTest {
```

Comment:
It would be good to add a test that updates the grace period through the subscribe handler, to ensure that code path keeps working.